### DISCUSSION TITLE
Invalid Parameters Error During Multiple Wells Model

### DISCUSSION THREAD

**Discussion by DataGeoranger** [pastas/pastas#395] [Q&A] [answered]
Hello!

My problem set is that I have one MW-oseries and six extraction wells and I am calibrating using Pastas based on the [Pastas example for multiple pumping wells ](https://pastas.readthedocs.io/en/latest/examples/010_multiple_wells.ipynb.html). There are two methods were suggested in the example. 1. create a model with a separate StressModel for each groundwater extraction. 2. Including the six extraction wells in one single WellModel. I tried both methods and have some questions about the results. Apologize for the lengthy descriptions and thank you in advance for the help!
Best,
Jack

**1.** When I tried to create separate StressModel for each extraction well, Pastas gave error said "is not a valid Parameters name" as shown _**the code below**_ . However I verified all the wells were added correctly by running `ml_h2.parameters`. **_Screen shot 1_** showed the parameters list after I add each stressmodel to the pastas model. The error happened when I run `ml_h2.solve(solver=ps.LmfitSolve)` as shown in **_screenshot 2_**. Could you please advise what could cause this error? 
```
#Create a model with a separate StressModel for each groundwater extraction
ml_h2 = ps.Model(oseries_h, name=oseries_h.name.replace(" ", "_"))
ml_h2.settings["freq"] = "H"
for wellstress in pumpingstresses:
    single_WM = ps.StressModel(wellstress, ps.Hantush, wellstress.name, up=False, meanstress=1)
    ml_h2.add_stressmodel(single_WM)
ml_h2.solve(solver=ps.LmfitSolve)
```

**Screen shot 1**
![image](https://user-images.githubusercontent.com/47303462/169735325-f4401650-2cfd-442e-a30b-b18e9496932f.png)

**Screen shot 2**
![image](https://user-images.githubusercontent.com/47303462/169735417-22bc8803-4ad5-44ee-8f58-973071b4e759.png)

**2.** When I solve six extraction wells in one single WellModel using **_the code below_**,  Pastas shows warning says "no distance passed HantushWellModel" as shown in **_screen shot 3_**. I tried to double check by using `ml_h3.stressmodels["Wells"].to_dict()`. The output (**_Screenshot 4 below_**) seems show that the distances have already got passed to the single WellModel. May I ask what could fix the warning and how could I verify that the distances got read into the WellModel correctly?

```
# WellModel - with Multiple Wells
distances = [1716.32,1358.17,290.85,3883.08,3912.90,3941.25]
ml_h3 = ps.Model(oseries_h, name=oseries_h.name.replace(" ", "_"))
mult_WM = ps.WellModel(pumpingstresses, ps.HantushWellModel, "Wells", distances=distances, up=False)
ml_h3.add_stressmodel(mult_WM)
ml_h3.settings["freq"] = "H"
ml_h3.solve(noise=True, solver=ps.LmfitSolve, 
            tmin = '2020-10-28 13:00:00', tmax= '2020-11-22 09:00:00', report=True)
```

**Screenshot 3**
![image](https://user-images.githubusercontent.com/47303462/169736083-867772e4-cf4d-4ea8-96a6-05c4e1ba2172.png)

**Screenshot 4**
![image](https://user-images.githubusercontent.com/47303462/169736538-9b7f47b1-56cf-49dc-a3a6-cdbc164357d5.png)

**3.** According to the [HantushWellModel descriptions](https://pastas.readthedocs.io/en/latest/concepts/hantush_response.ipynb.html) - **_screenshot 5_** below, is the parameter **A'** represent the transmissivity of the aquifer in HantushWellModel? For Hantush Model, there is a parameter (**Ko**) for parameter **A**, is **Ko** represent the hydraulic conductivity in the equation?

**Screenshot 5**
![image](https://user-images.githubusercontent.com/47303462/169738645-fcfa7f77-8116-4f43-ac51-4b080b030fe1.png)

**Comment by dbrakenhoff** [comment]
Hi Jack,

Don't worry about the lengthy descriptions, this helps us understand your problem more quickly! 

1. I think the issue here is that LmFit doesn't like the parameter names starting with a numeric character. Similar to Python's requirements in naming variables you're not allowed to start a name with a number. Can you try renaming the wells to start with a letter?
2. This is a slightly misleading warning. WellModel takes both 3 parameters or 4 parameters (when the distance to a specific well is included). When 3 parameters are passed it assumes the distance `r=1`. Whenever only 3 parameters are passed to `WellModel.simulate()` the logger throws out a message that `r=1`, in case that's not what you intended. In your case, this is being caused by the fit_report. Part of the fit report is a check whether the response time t_95% is smaller than the length of the timeseries. For that check `r` is assumed to be 1, which is generating that message. So nothing to be worried about, but I understand that it makes you question whether you did everything correctly. I'll think about whether we can improve the message or change when it is logged.
3. `K0` the modified Bessel function of the second kind, order 0 (you can get it from scipy: `from scipy.special import k0`). I recommend checking the [response functions notebook](https://github.com/pastas/pastas/blob/dev/concepts/response_functions.ipynb) towards the bottom where the Hantush response function parameters are related to geohydrological parameters.

Hope that answers your questions! If something is unclear, let me know!

Kind regards,

Davíd

**Reply by DataGeoranger** [reply]
Thank you very much David. Please see below some follow-up questions.

> I think the issue here is that LmFit doesn't like the parameter names starting with a numeric character. Similar to Python's requirements in naming variables you're not allowed to start a name with a number. Can you try renaming the wells to start with a letter?
>
**1.** Rename well names with a letter solved the error! But the calibration plot (**screenshot 1**) using method - "**create separate StressModel for each extraction well**" was not as good (not even close) as calibration plot results using method - **WellModel with multiple pumping wells** (**screenshot 2**). Both were solved at frequency of 1 day. Have you observed similar difference in calibration performance between method 1 and method 2 for multiple extraction wells?  I wonder if this is due to there are too many initial parameters to change in method 1. 

**Screenshot 1**
![image](https://user-images.githubusercontent.com/47303462/169886350-4cf412e1-2ff1-41b5-ad31-304d22a81021.png)

**Screenshot 2**
![image](https://user-images.githubusercontent.com/47303462/169886692-d6a67264-4655-43e6-ab9c-e65de5bfa673.png)

> K0 the modified Bessel function of the second kind, order 0 (you can get it from scipy: from scipy.special import k0). I recommend checking the [response functions notebook](https://github.com/pastas/pastas/blob/dev/concepts/response_functions.ipynb) towards the bottom where the Hantush response function parameters are related to geohydrological parameters.

Thanks! The notebook was very helpful. 
**2a** Is it correct to interpret that for WellModel with multiple pumping wells, the `gain = AKo(2rb**0.5)` represents the drawdown caused by pumping at 1 unit pumping rate - cubic feet per day from each well? 
**2b**. Is there a physical meaning of parameter `A` as shown in **screenshot 3**? The equation for `A = (1/2*pi*T)` has a parameter `T` and I wonder if `T` would represent the transmissivity of the aquifer, like it was suggested in the tutorial (screenshot 4). I am trying to understand if Pastas allow me to estimate hydraulic conductivity/transmissivity from any of the calibrated parameters in the WellModel to compare to the field testing results from a pumping test at my Site.

**Screenshot 3**
![image](https://user-images.githubusercontent.com/47303462/169890269-d9932ff2-bd5b-4572-8dae-e3d994d2dfc0.png)

**Screenshot 4**
![image](https://user-images.githubusercontent.com/47303462/169904450-b13a5d42-366a-4e4f-a413-1deb9510d257.png)

**3.** For WellModel with multiple pumping wells, I also tried to compare the calibration results  when running models at `freq = 'D'` and at `freq = 'H'`. Because the fitting statistic (R squared = 97.90%) was good when I solve the model at `freq = 'D'`, but the calibrated value plot missing some sharp change in water level as shown in the oseries (oseries has a frequency at 'H'), please see **screenshot 5** curve circled in blue. But very interestingly, the calibration statistics (R squared = 90.43%) became worse when I solve the model at `freq = 'H'` instead of capturing the observed head change as shown in **screenshot 6** - circled in blue. I also set my initial parameter value to be the same as the optimize parameters from when model was solved at `freq='D'`.  Could you please advise what else I can do to improve the matching at places where a sharp change observed in the water levels? Should I go finer time frequency (my original transducer data was at every 10 mins)

**Screenshot 5**
![image](https://user-images.githubusercontent.com/47303462/169891505-75a8b98c-66aa-4960-83a5-6716659f2914.png)

**Screenshot 6**
![image](https://user-images.githubusercontent.com/47303462/169891828-9f9336c4-5163-4392-a41a-cb2ef41c06e7.png)

**4.** Looking at the step responses plot (lower right panel) in **screenshot 5** and **screenshot 6** as shown above, it seems the x axis interval has changed when going from solving model at freq = 'D' to solving model at freq = 'H'. Could this be a labeling error? Because I am thinking the step response time wouldn't change when solving at different freq for the same pumping stresses, however, the resulted response plots for freq = 'D' shows it reached steady state after 1 day, but for freq = 'H', it reached steady state 0.4 days.

**Comment by dbrakenhoff** [comment, marked as answer]
Hi Jack,

1. In theory the model with multiple stressmodels (one for each well) should be able to outperform the model using a WellModel as it has more freedom to fit the observations. Have you tried solving the model without noise? Also try solving with `ps.LeastSquares` as the solver. In my experience LeastSquares is sometimes better in models with separate stressmodels, whereas `ps.LmfitSolve` is better for models with a WellModel (as it probably handles the small parameter values of `b` better). Also the initial values of the parameters can influence the optimization, so playing around with those could also help.
2. a) Correct. b) T is the transmissivity, c is the resistance of the overlying aquitard, S is the storage coefficient. Not that it is only appropriate to relate the optimized Pastas parameters to these geohydrological parameters if your subsurface is similar to the conceptual model in  Hantush's equation: a single semi-confined aquifer. 
3. And yes, matching those moments when pumping starts is often a challenge. There are only a few observations in the steep section of the drawdown plot, which means that the optimization often tends to focus more on the steady-state periods. I don't have a great solution ready, but you could play around with the weights assigned to the observations to give the early period after start of pumping more relative importance? You can do this by passing weights to `ml.solve(weights=...)`. Other ideas are to use the 10T frequency and keep all observations in the steep sections and drop observations in the more steady-state periods. I would also check your pumping discharge timeseries as it seems to start and end too soon relative to the observations?
4. One reason the step response can change between `freq="H"` and `freq="D"` is if the units of the timeseries change. So if the pumping volume is divided by 24 to get the hourly volumes that would also lead to a change in the step response function, but then you'd expect a factor 24 difference. Another reason can be that the models have a different result because the model fits are different. Ideally, you'd want to get the same result, but that will only occur if the units are the same and the model fit is similar. Another thing is the step response is shown in steps of `1D` when the `freq="D"` so you're not seeing the step response in a very high resolution, it could be that the steady state is also reached in about 0.4 days in this case but you're not seeing that in the plot. I would compare these models (especially the parameters, making sure the units are consistent) to check if the fit is different.

Hope that answers your questions somewhat?

**Reply by DataGeoranger** [reply]
Thank you again! David. Please see below my follow up questions and thanks so much for your patience with me.

> 1. Also try solving with `ps.LeastSquares` as the solver.

**1.** I tried to solve it with `ps.LeastSquares` with `noise=False`, that helps, but not satisfactory compared to a WellModel (**Screenshot 1**). Then I also tried assign better initial value of the parameters (based on results from a WellModel) for one well, the calibration result (R squared) was improved from from 78% to 80%.(**Screenshot 2**) So I went all the way and asssigned same initial values of parameters to each extraction well (I have six extraction wells in total), then I finally get similar calibration results as comparable to a WellModel (**Screenshot 3**). For which the result are satisfactory, but makes it difficult in real practice using method 1. I wonder if there is a way to hand calculate the initial values for parameters for `A, a, and b`. Could you please shed some light on what site-specific parameters can be used to estimate initial parameter values that Pastas required?

**Screenshot 1**
![image](https://user-images.githubusercontent.com/47303462/170155667-ab970ae5-a043-40fe-9992-efee8cf2f79b.png)
**Screenshot 2**
![image](https://user-images.githubusercontent.com/47303462/170156261-51c57d26-6d1c-460e-b884-e23142c8a40c.png)
**Screenshot 3**
![image](https://user-images.githubusercontent.com/47303462/170156587-a93f2b7d-8cae-44c5-933f-ca2e4ef25a45.png)

> 2\. a) Correct.

**2a.** I was able to extract the gain for each of my extraction well (**Screenshot 4**) by using code 
```
for i in range(len(well_list2)):
    p = wm1.get_parameters(model=ml_h3, istress = i)
    gain = wm1.rfunc.gain(p). 
```
If I divide the gain for each well by `parameter - Ko, distance, and b` according to the equation `gain = AKo(2rb**0.5)`. Would the resulted A represent the transmissivity between a particular extraction well and the interested monitoring well? And would the gain represented the level of influence from each particular extraction well to the interested monitoring well? i.e., the higher the gain number, the higher the influence. 
**Screenshot 4**
![image](https://user-images.githubusercontent.com/47303462/170165947-6a8da093-5735-4c40-ae61-c778d7036254.png)

> 2\. b) T is the transmissivity, c is the resistance of the overlying aquitard, S is the storage coefficient. Not that it is only appropriate to relate the optimized Pastas parameters to these geohydrological parameters if your subsurface is similar to the conceptual model in  Hantush's equation: a single semi-confined aquifer.

**2b.** I extracted the parameters from WellModel using `ml_h3.parameters` and the result are in **Screenshot 5**. Based on your explanation, I could calculate `T = 1/(2*pi*A) as 1/(2*3.14*-0.001624)`, which worked to be about `-98.0`. My question is what would be the unit of the number if I solved my Pastas model at` freq = 'H'`. Would it be `feet square per hour`? My other question is that what would be the physical meaning if the estimated `transmissivity of T` is negative, maybe I am missing some negative sign from the equation?
For estimating storage from equation a = cS in Pastas Hantush function, it seems that the parameter **c and S** are correlated. I wonder if there is empirical literature value for `c` (resistance of the overlying aquitard) that would allow me to calculate storage (S)?

**Screenshot 5**
![image](https://user-images.githubusercontent.com/47303462/170167124-29b49e6c-9afc-451b-b022-fe2129b5acfa.png)

**3.** I was also able to extract the gain value for each extraction well from the model (create separate StressModel for each extraction well) and compared to the gain for each extraction well from one WellModel as shown in **Screenshot 6**. It looks like there are some differences in estimated gain value between the two methods as shown in **Screenshot 7** and **Screenshot 8**. They were both solved at same freq and no noise effect, and fitting results from both looks pretty decent. I wonder the estimate gain from which model would be more reliable? Maybe the separate StressModel gain values are less reliable given there are many parameters the solver had to estimate?

**Screenshot 6**
![image](https://user-images.githubusercontent.com/47303462/170170896-a0ba7bf2-f07f-4b2b-9d7f-efbf6f43ed7d.png)
**Screenshot 7 - separate stress for each extraction well**
![image](https://user-images.githubusercontent.com/47303462/170170293-4fbfe56a-7677-42aa-9fb2-88ccfbc72725.png)
**Screenshot 8- one WellModel for multiple extraction wells** 
![image](https://user-images.githubusercontent.com/47303462/170170493-b1a9b07f-f514-4162-9485-2946fdfcf583.png)

> 3\. You can do this by passing weights to `ml.solve(weights=...)`

**4.** Assign weight helped compared before, it improve my R square from 90% to 96% when I assigned weights and solve at `freq = 10T`, and I can observe the improvements from the curve as well! (**Screenshot 9**). I maybe need to further play with the weights a bit more. Thanks for the suggestions!
**Screenshot 9**
![image](https://user-images.githubusercontent.com/47303462/170176717-cf46c77e-0958-4dae-928d-fc9902f0bdf5.png)

**Reply by dbrakenhoff** [reply]
Hi Jack,

1. The estimation of initial parameters is a topic of ongoing research. We implemented some methods in pastas but these estimates are by no means holy, and as you've seen, altering the starting parameters can improve the result. As for estimating the parameters. If you're using Hantush and you know something about the subsurface, you can use those estimates of subsurface parameters (transmissivity, storage coefficient, aquitard resistance) to get some estimates. Another method you could try is to optimize your model in two steps, the first time without the noisemodel, and the second time using the optimal parameters from the first estimate:

```python
ml.solve(noise=False)
ml.sovle(noise=True, initial=False)
```
But having said that, in your case it seems as if there is insufficient information in the timeseries to be able to quantify the influence of most of the wells. In Screenshots 1 and 2 it is obvious that some wells are being used to simulate trends (even trends cancelling one another out). In the final screenshot, only one well is relevant. 

2. a) For WellModel transmissivity `T = -1 / (2 * np.pi * wells_A)`, for individual stressmodels with Hantush response `T = -k0(2 * np.sqrt(b)) / (2 * np.pi * wells_A)`.

> Would the resulted A represent the transmissivity between a particular extraction well and the interested monitoring well?
The resulting transmissivity is an estimate of the transmissivity of the aquifer, if it was a single semi-confined aquifer.
> And would the gain represented the level of influence from each particular extraction well to the interested monitoring well? i.e., the higher the gain number, the higher the influence.
The gain for the wells represents the steady-state drawdown at the observation point for one unit of pumping for that particular well, i.e. the drawdown from well_X in feet per 1 cubic foot per day. It can be useful to scale the pumping discharge to larger numbers (I don't know what imperial logic would suggest, but e.g. 1000 cubic feet per day).

b) The transmissivity should be positive, so yea, you need a minus sign in there (probably the same sign as your pumping discharge timeseries). As for `cS`, you're correct that you cannot identify these individually. You'll need to find some estimate for one of the two to be able to get a value for the other one. As for literature on aquitard resistance, there should be plenty lying around (perhaps from pumping tests, or drilling samples and literature values for vertical hydraulic conductivities for specific soil types). I can't really help you there beyond that. 
As for the unit of the parameter A when solving for `freq="H"`, it depends on the timeseries settings. If you're using divide to resample, then it would be hours. If you're using bfill, then the units should remain the same. 

3. Yea, the models are quite different I think, so it's not strange that the estimated gains are different between the two models. You can plot the estimated gain vs. the distance to check if that provides a logical drawdown curve. In the case of the WellModel this will be the case as it forces this logic on the model, in the individual case, you'll probably see that this isn't the case, which would in my opinion be a good reason to reject the model. And yea the high number of parameters can be a challenge for the model, especially if there isn't sufficient information in the timeseries to accurately estimate all of them. 

4. Nice!

**Reply by DataGeoranger** [reply]
Thank you very much! David. Please see my follow up response below. 

> But having said that, in your case it seems as if there is insufficient information in the timeseries to be able to quantify the influence of most of the wells.
1. Follow your thought of insufficient information, I agree that I might've I overfit the observed WL in the MW by adding too many stresses (pumping). So I did some more test by running WellModel with just one well, two wells, three wells, and four wells (calibrate to the same MW at same freq). The results shows that the model fit was pretty good when I just had one pumping well (4555), adding extra wells didn't bring too much benefit in terms of calibration as shown in **screenshot 1, screenshot 2, screenshot 3, and screenshot 4**. And the gain value became smaller and smaller as I added more wells that are further away (**Screenshot 4 Table**). **Screenshot 5** shows the spatial locations of the MW and four extraction wells.

As you can see, my MW locates in the midst of several nearby EWs, and would likely be impacted by those extraction wells that I added. However, it seems Pastas suggest only one EW would be enough to explain observed WLs in the MW. My questions then are:

a), Is the result in which gain reduced significantly as I added additional EWs due to the mathematical formulation of the WellModel that when the distance of a EW is further away from the MW, the gain will reduce, or the gain reduction actually reflects the physical phenomenon that the level of hydraulic connection between the added EW and MW becomes weaker?

b), There are actually more wells at my site. Is there any criteria I could follow to decide how many more EWs I should add to the model? Maybe based on the site hydraulic testing results and analytical solution to calculate distance of drawdown?
Or in this case, one EW (4555) is just enough because it fit the MW pretty good?

Screenshot 1
![image](https://user-images.githubusercontent.com/47303462/170409265-e9ccc592-50a6-46ff-a6e7-e873a3f6ca15.png)

Screenshot 2
![image](https://user-images.githubusercontent.com/47303462/170409308-2186919a-2f66-4cb5-92f2-09fbc62c134c.png)

Screenshot 3
![image](https://user-images.githubusercontent.com/47303462/170409349-3114aec4-c67e-497b-a216-cf4e135284e7.png)

Screenshot 4
![image](https://user-images.githubusercontent.com/47303462/170409385-d5531093-2ea9-4227-a2e3-b032d9f840b2.png)

Screenshot 5
![image](https://user-images.githubusercontent.com/47303462/170409687-3bea991d-12bf-4288-b415-1f70c7530c34.png)

2. To further explore the uncertainty of the gain value I got for each well, I used the tutorial example to calculate the variance of the gain. As you can see in **screenshot 6**, the variance gain is also becomes smaller and smaller for those EWs that are further away. I guess it is because of distance. But intuitively I would expect the variance gain increases, as uncertainty of estimating gain for those wells becomes higher and higher due to insufficient information from the MW water levels, such that std and variance should be bigger?
**Screenshot 6**
![image](https://user-images.githubusercontent.com/47303462/170411866-759535d4-10ef-43ad-99e8-63d91557c315.png)


3.
> 3\. Yea, the models are quite different I think, so it's not strange that the estimated gains are different between the two models. You can plot the estimated gain vs. the distance to check if that provides a logical drawdown curve. In the case of the WellModel this will be the case as it forces this logic on the model

I am not very sure how to extract the results based on the suggestion(maybe I already have the results?sorry). Could you please provide a bit more instruction or code example to visualize this gain vs. distance curve? Thanks!

4. At the same site and same well, I tried estimate the long-term yearly recharge flux using the non-linear model (Flex) using rainfall data nd ETo data in the past 20 years. Can I post related questions here or you prefer me to start a new discussion for the recharge rate estimation questions?

Thanks again for the suggestions and help!

**Reply by dbrakenhoff** [reply]
Hi Jack,

It was nice meeting in you in person! I was running through the issues and discussions and saw there were still some unanswered questions. 

First though, I wanted to say that I think that you're running into the limitations of having only a relatively short timeseries to model using timeseries analysis, which might make it difficult to estimate the individual contributions of wells.

Can you add a screenshot of your head observations and the pumping stresses with shared time axes? That way I can see how different the pumping rates are for each well, and how they relate to the total period with head observations.

1. a) This is difficult to answer, but the gain getting smaller for the same well as you add extra pumping stresses could be the result of some of the drawdown being attributed to wells farther away? One note also, since your gain is so small, I would recommend changing the units of your pumping stresses to a different unit, i.e. Mm^3/year instead of m^3/day for example. That way the gain will represent the drawdown per 1 Mm3/yr of pumping, leading to a larger number in the results. Tiny parameter values might also be challenging to optimize for the solver.

b) A simple analysis of the characteristic length (\lambda = \sqrt{Tc}) for your aquifer could shed some insight? Also just trying to add wells to your timeseries model could give some insight into whether it improves as you add more wells, and whether those wells have a significant effect?

2. I'm not sure what makes sense here, the variance of the gain for each well is based on the same formula and only scaled by `r` so I guess it makes sense it gets smaller as `A` also get smaller? Not sure how the ratio between the two changes though. 

3. Calculate the gain using `wm.rfunc.gain()` and just extract the distances from `wm.distances` and plot them. For the gain:

```python
for i in range(len(well_list2)):
    p = wm1.get_parameters(model=ml_h3, istress = i)
    gain = wm1.rfunc.gain(p). 
```

4. This is a separate question ;).

Cheers,

Davíd
